### PR TITLE
Docker refresh

### PR DESCRIPTION
**Pull request by cypof**
I'd like to have an [official image for Caffe](https://hub.docker.com/r/bvlc/caffe), so updated the builds to the latest Ubuntu, CUDA and cuDNN, and added NCCL. Also I'm not sure I see the point of maintaining the docker Makefile and templates now that images build automatically, so I tentatively removed them.

### CODE REVIEW

File: docker/cpu/Dockerfile [hunk 1]
```diff
@@ -20,17 +20,19 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
         python-scipy && \
     rm -rf /var/lib/apt/lists/*
 
 ENV CAFFE_ROOT=/opt/caffe
 WORKDIR $CAFFE_ROOT
 
-# FIXME: clone a specific git tag and use ARG instead of ENV once DockerHub supports this.
-ENV CLONE_TAG=master
+# FIXME: use ARG instead of ENV once DockerHub supports this
+ENV CLONE_TAG=1.0.0-rc4
```
Comment (flx42) [inline docker/cpu/Dockerfile]: That's not a valid bvlc/caffe tag, right?

File: docker/cpu/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:14.04
+FROM ubuntu:16.04
 MAINTAINER caffe-maint@googlegroups.com
```
Comment (flx42) [inline docker/cpu/Dockerfile]: `MAINTAINER` is being [deprecated](https://github.com/docker/docker/pull/25466) in Docker 1.13, you should use a label instead, might as well change it now.
You can use something like [this](https://github.com/NVIDIA/nvidia-docker/blob/ef3e6055059b515e8e3ee0801848bbad8b5efde9/centos-7/cuda/7.5/devel/Dockerfile#L2) instead.

File: docker/gpu/Dockerfile [hunk 3]
```diff
@@ -20,19 +20,22 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
         python-scipy && \
     rm -rf /var/lib/apt/lists/*
 
 ENV CAFFE_ROOT=/opt/caffe
 WORKDIR $CAFFE_ROOT
 
-# FIXME: clone a specific git tag and use ARG instead of ENV once DockerHub supports this.
-ENV CLONE_TAG=master
+# FIXME: use ARG instead of ENV once DockerHub supports this
+ENV CLONE_TAG=1.0.0-rc4
```
Comment (flx42) [inline docker/gpu/Dockerfile]: Ditto.

File: docker/gpu/Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM nvidia/cuda:7.5-cudnn5-devel-ubuntu14.04
+FROM nvidia/cuda:8.0-cudnn5-devel-ubuntu16.04
 MAINTAINER caffe-maint@googlegroups.com
```
Comment (flx42) [inline docker/gpu/Dockerfile]: Ditto.

File: docker/gpu/Dockerfile [hunk 5]
```diff
@@ -20,19 +20,22 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
-# FIXME: clone a specific git tag and use ARG instead of ENV once DockerHub supports this.
-ENV CLONE_TAG=master
+# FIXME: use ARG instead of ENV once DockerHub supports this
+ENV CLONE_TAG=1.0.0-rc4
 
 RUN git clone -b ${CLONE_TAG} --depth 1 https://github.com/BVLC/caffe.git . && \
-    for req in $(cat python/requirements.txt) pydot; do pip install $req; done && \
+    pip install --upgrade pip && \
+    cd python && for req in $(cat requirements.txt) pydot; do pip install $req; done && cd .. && \
+    git clone https://github.com/NVIDIA/nccl.git && cd nccl && make -j install && cd .. && \
```
Comment (flx42) [inline docker/gpu/Dockerfile]: You should probably remove the NCCL directory after install, to save space.

**Comment by cypof** [comment]
Thanks @flx42 should be good now

**Comment by flx42** [comment]
@cypof what about the `git clone` tag?

**Comment by cypof** [comment]
I pushed a tag, it should be there

**Comment by flx42** [comment]
Great, LGTM then.

**Comment by flx42** [comment]
Actually, not LGTM. `-DUSE_NCCL=1` is not supported in current BVLC/caffe. I take it this PR comes after #4563?

**Comment by cypof** [comment]
Ah yes, I was assuming the nccl would make it. I just refreshed #4563 history so it should be ready now. Let's wait until it goes in.

**Comment by flx42** [comment]
Now that #4563 was merged, you should create a new tag (`rc5`?) for your Dockerfile.

**Comment by shelhamer** [comment]
Dropping the Makefile seems fine as long as the docs clearly explain how to build the images yourself for those who want to work with their own branches.

**Comment by cypof** [comment]
Yes this PR improves the readme too. It's simpler to get started and it shows how to build the images. I think this is good to go, I will make sure to update the tag as soon as we have an official rc4.

**Comment by cypof** [comment]
Pointing to rc4 now.

**Comment by flx42** [comment]
LGTM

**Comment by shelhamer** [comment]
@cypof feel free to merge when ready

**Pull request merged**